src/gt4py/storage/allocators.py

```diff
@@ -181,14 +188,113 @@ def allocate(
         ...
 
 
-@dataclasses.dataclass(frozen=True, init=False)
-class _BaseNDArrayBufferAllocator(abc.ABC, Generic[core_defs.DeviceTypeT]):
-    """Base class for buffer allocators using NumPy-like modules."""
+class MemoryResourceHandler(Protocol[core_defs.DeviceTypeLiteralT]):
```

Comment:
After reading more in this file, my suggestion doesn't make sense for what this is intended to do, but I am not sure I like this grouping into a class with classmethods.

The trigger for this change was `byte_bounds`, the following solutions looks cleaner to me:
- `malloc` returns buffer and address, because that's the only place we use it (if I didn't miss anything)
- `address_of` as single dispatch and we register cupy and numpy buffer

`tensorize` is duplicated, it seems cleaner to use a free function and pass the namespace

Reply:
While looking at this I am trying to understand at which abstraction level would we add a low level pool allocator.